File: java/src/test/java/finance/clarian/sdk/MockTransport.java
package finance.clarian.sdk;

import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

/** Fake HTTP transport for unit tests (no network). */
final class MockTransport implements HttpTransport {

    record Recorded(
            String method,
            String url,
            String path,
            Map<String, String> headers,
            byte[] body
    ) {
        String bodyUtf8() {
            return body == null ? "" : new String(body, StandardCharsets.UTF_8);
        }
    }

    private final Function<Recorded, Response> handler;
    final List<Recorded> calls = new ArrayList<>();

    MockTransport(Function<Recorded, Response> handler) {
        this.handler = handler;
    }

    static Response json(int status, Object payload) {
        if (payload == null) {
            return new Response(status, new byte[0]);
        }
        if (payload instanceof byte[] bytes) {
            return new Response(status, bytes);
        }
        if (payload instanceof String s) {
            return new Response(status, s.getBytes(StandardCharsets.UTF_8));
        }
        return new Response(status, Json.encode(payload).getBytes(StandardCharsets.UTF_8));
    }

    Recorded last() {
        if (calls.isEmpty()) {
            throw new AssertionError("no requests recorded");
        }
        return calls.get(calls.size() - 1);
    }

    @Override
    public Response exchange(
            String method,
            String url,
            Map<String, String> headers,
            byte[] body,
            Duration timeout
    ) {
        String path = URI.create(url).getRawPath();
        String query = URI.create(url).getRawQuery();
        if (query != null && !query.isEmpty()) {
            path = path + "?" + query;
        }
        Recorded rec = new Recorded(
                method,
                url,
                path,
                new LinkedHashMap<>(headers),
                body == null ? new byte[0] : body
        );
        calls.add(rec);
        return handler.apply(rec);
    }
}

File: python/tests/mock_transport.py
"""In-memory HTTP transport for unit tests (no network)."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable, Mapping, Optional
from urllib.parse import urlparse


@dataclass
class RecordedRequest:
    method: str
    url: str
    path: str
    headers: dict[str, str]
    body: Optional[bytes]


Handler = Callable[[RecordedRequest], tuple[int, bytes | str | dict[str, Any]]]


@dataclass
class MockTransport:
    """Records requests and returns canned responses via a handler."""

    handler: Handler
    calls: list[RecordedRequest] = field(default_factory=list)

    def __call__(
        self,
        method: str,
        url: str,
        headers: Mapping[str, str],
        body: Optional[bytes],
        timeout: float,
    ) -> tuple[int, bytes]:
        parsed = urlparse(url)
        path = parsed.path
        if parsed.query:
            path = f"{path}?{parsed.query}"
        # When base_url is a full gateway URL, path includes the gateway prefix.
        # Tests pass base_url like "https://mock" so path is just the API path.
        req = RecordedRequest(
            method=method,
            url=url,
            path=path,
            headers=dict(headers),
            body=body,
        )
        self.calls.append(req)
        status, payload = self.handler(req)
        if isinstance(payload, bytes):
            raw = payload
        elif isinstance(payload, str):
            raw = payload.encode("utf-8")
        else:
            raw = json.dumps(payload).encode("utf-8")
        return status, raw

    @property
    def last(self) -> RecordedRequest:
        if not self.calls:
            raise AssertionError("no requests recorded")
        return self.calls[-1]
